### PR TITLE
Add failing test case for decimal.rounded case to log based

### PR DESCRIPTION
**Pull request by bhtowles**
# Description of change
Added decimal data type to the log based replication test based on TDL-16037, and TDL-16136.  Found additional edge cases were still hitting the error.  Generated TDL-19094 to resolve remaining error cases.

# Manual QA steps
None
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_dynamodb_log_based.py [hunk 1]
```diff
@@ -27,6 +27,8 @@ def generate_items(self, num_items, start_key=0):
             record = {
                 'int_id': i,
                 'string_field': self.random_string_generator(),
+                # BUG https://jira.talendforge.org/browse/TDL-19094
```
Comment (kspeer825) [inline tests/test_dynamodb_log_based.py:30]: Has this issue been pointed?

File: tests/base.py [hunk 2]
```diff
@@ -148,6 +149,60 @@ def clear_tables(self, table_names=ALL_TABLE_NAMES_TO_CLEAR):
     def random_string_generator(size=6, chars=string.ascii_uppercase + string.digits):
         return ''.join(random.choice(chars) for x in range(size))
 
+    @staticmethod
+    def random_decimal_generator(digits=string.digits):
```
Comment (kspeer825) [inline tests/base.py:153]: We have control over the data here right, so instead of randomizing the value can we hardcode a few different values that cover each test case? Like have a new row for each case.

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/test_dynamodb_log_based.py (hunk 1 above)
Comment (bhtowles) [reply tests/test_dynamodb_log_based.py:30]: No, do I need to give it a review label?

File: tests/base.py (hunk 2 above)
Comment (bhtowles) [reply tests/base.py:153]: Yes we do, but I'm not sure we know what all the problem test cases are.  Given the existing cases start with 100 rows it seemed good to randomize to look for additional problem cases.

File: tests/test_dynamodb_log_based.py (hunk 1 above)
Comment (kspeer825) [reply tests/test_dynamodb_log_based.py:30]: yep!

File: tests/base.py (hunk 2 above)
Comment (kspeer825) [reply tests/base.py:153]: It looks like you identified some boundaries in your randomization. I think testing the max min, zero and something in between those boundaries would be sufficient. But I guess there are multiple boundaries to consider. Max/min value, max/min precision, min/max scale 🤯 . The hp-mysql and hp-postgres tests have examples for this.

File: tests/test_dynamodb_log_based.py (hunk 1 above)
Comment (bhtowles) [reply tests/test_dynamodb_log_based.py:30]: Done!

File: tests/base.py (hunk 2 above)
Comment (bhtowles) [reply tests/base.py:153]: Agreed, I'm also not sure how different formatting is handled and if signed or unsigned has significant impact.  Are you suggesting a refactor to align with the hp-mysql and hp-postgres approaches or are we good to leave it random to flush out all of these over time?  To be fair min precision is not currently covered.

Comment (kspeer825) [reply tests/base.py:153]: I think it should be refactored to hardcode values for the different partitions you're covering. The random approach is a good topic to get @HarrisonMarcRose opinion on. But I think it made sense for field selection with google-ads because their weren't clear partitions in the available fields, where we could confidently say if x works so does y. But there are almost always clear partitions in numeric datatypes. And in most datatypes I think.

Comment (bhtowles) [reply tests/base.py:154]: Refactored to a static list of 20 boundry-ish values that are cycled through by the generator.

File: tests/base.py [hunk 3]
```diff
@@ -148,6 +150,43 @@ def clear_tables(self, table_names=ALL_TABLE_NAMES_TO_CLEAR):
     def random_string_generator(size=6, chars=string.ascii_uppercase + string.digits):
         return ''.join(random.choice(chars) for x in range(size))
 
+    #@staticmethod # TODO is this needed?
+    def random_decimal_generator(self):
+        # Refactor per PR review
+        # List boundry cases (20)
+        # max exponent with max precision and max digit magnitute 0.999...e126 (pos and neg)
+        # min exponent with max precision and max digit magnitute 0.999...e-128 (pos and neg)
+        # TODO card out test overflow of above two cases?
```
Comment (kspeer825) [inline tests/base.py:159]: Could you link the card for this  get linked here so future testers can search for the ticket id to reproduce this issue

File: tests/base.py [hunk 4]
```diff
@@ -148,6 +150,43 @@ def clear_tables(self, table_names=ALL_TABLE_NAMES_TO_CLEAR):
     def random_string_generator(size=6, chars=string.ascii_uppercase + string.digits):
         return ''.join(random.choice(chars) for x in range(size))
 
+    #@staticmethod # TODO is this needed?
+    def random_decimal_generator(self):
```
[the remark this answers is not in the document]
Comment (bhtowles) [reply tests/base.py:154]: The decimal_index variable is considered a class variable in this sense then?  If so I'll delete the comment.

File: tests/base.py (hunk 3 above)
Comment (bhtowles) [reply tests/base.py:159]: Let's discuss.  Are you referring to the decimal.Rounded card or a card specific to the precision 38 with e-128?

Comment (kspeer825) [reply tests/base.py]: the ticket is linked now resolving

File: tests/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kspeer825) [reply tests/base.py]: yes, looks good

**Review by kspeer825 — APPROVED** [review]
